### PR TITLE
Patient creation

### PR DESCRIPTION
**Pull request by marcovilla86**

### CODE REVIEW

File: src/main/java/ca/uhn/fhir/jpa/tcp/interceptor/TokenAuthorizationInterceptor.java [hunk 1]
```diff
@@ -1,93 +1,92 @@
...
-				X509EncodedKeySpec keySpec = new X509EncodedKeySpec(buffer);
-				RSAPublicKey publicKey = (RSAPublicKey) keyFactory.generatePublic(keySpec);
+      String authHeader = theRequestDetails.getHeader("Authorization");
+      if (authHeader == null) {
+        throw new AuthenticationException("Missing authorization header");
+      }
+      String inum = "";
+      try {
+        String token = authHeader.split(" ")[1];
+        DecodedJWT jwt = JWT.decode(token);
```
Comment (jdkizer9) [inline src/main/java/ca/uhn/fhir/jpa/tcp/interceptor/TokenAuthorizationInterceptor.java:66]: Is this verifying the token? Per the source code [here](https://github.com/auth0/java-jwt/blob/dd6bc25d9a7c2857bae29757a9198b798ac5734f/lib/src/main/java/com/auth0/jwt/JWT.java#L44), it seems to just be decoding the token w/o doing any verification. I think we need to follow the instructions [here](https://github.com/auth0/java-jwt#verify-a-token).

**Review by jdkizer9 — COMMENTED** [review]
Left one comment, but generally, I don't feel as though I don't have enough context to provide a good review.

File: src/main/java/ca/uhn/fhir/jpa/tcp/interceptor/TokenAuthorizationInterceptor.java (hunk 1 above)
Comment (marcovilla86) [reply src/main/java/ca/uhn/fhir/jpa/tcp/interceptor/TokenAuthorizationInterceptor.java:66]: I mentioned before that we've been having troubles with the verification after moving to the new gluu version, Raul was helping me with this, now that I've finished the patient creation stuff I can do more research in order to fix the signature verification

Comment (marcovilla86) [reply src/main/java/ca/uhn/fhir/jpa/tcp/interceptor/TokenAuthorizationInterceptor.java:66]: as you can see on the previous version I was verifying the signature as you are suggesting, working on it again..

**Review by misterzero — APPROVED** [review]
The changes look good, but I think this is significant enough to warrant adding unit tests. I see there are some tests in the original project if you want to rebase to include them. Let me know if you need guidance with this.

I'm still approving the PR once any other comments are addressed, because we need to deploy this feature to unblock other work and I realize the testing may take some time. I will create a new ticket to address unit testing.

File: src/main/java/ca/uhn/fhir/jpa/tcp/interceptor/TokenAuthorizationInterceptor.java [hunk 2]
```diff
@@ -1,93 +1,92 @@
...
-				 * }
-				 **/
-			} catch (JWTVerificationException | NoSuchAlgorithmException | InvalidKeySpecException exception) {
-				// Invalid signature/claims
-				throw new AuthenticationException(exception.getMessage());
-			}
-		}
-		return new RuleBuilder().allowAll().build();
-	}
+      return super.applyRulesAndReturnDecision(theOperation, theRequestDetails, theInputResource, resourceId,
```
Comment (jdkizer9) [inline src/main/java/ca/uhn/fhir/jpa/tcp/interceptor/TokenAuthorizationInterceptor.java:77]: Do we handle permission checking based on scope? How do we ensure that the resource is "owned" by the user associated with the token?

File: src/main/java/ca/uhn/fhir/jpa/tcp/interceptor/TokenAuthorizationInterceptor.java (hunk 1 above)
Comment (marcovilla86) [reply src/main/java/ca/uhn/fhir/jpa/tcp/interceptor/TokenAuthorizationInterceptor.java:73]: I've added the signature verif functionality 
https://github.com/the-commons-project/hapi-fhir-jpaserver-starter/pull/8/files#diff-db8eda07481b3f00c0a2f8a6ee065407baf2f74ec1c8a02e3588a27405d34650R74-R78

File: src/main/java/ca/uhn/fhir/jpa/tcp/interceptor/TokenAuthorizationInterceptor.java (hunk 2 above)
Comment (marcovilla86) [reply src/main/java/ca/uhn/fhir/jpa/tcp/interceptor/TokenAuthorizationInterceptor.java:88]: when creating a patient we need to trust the token only, on subsequence calls now I'm validating that the patient id belongs to the linked resources on empi

https://github.com/the-commons-project/hapi-fhir-jpaserver-starter/pull/8/files#diff-b991193457013b3c05a76d754be6bf2ee7db702ca5cfd879784a63073bb86f96R64-R67

https://github.com/the-commons-project/hapi-fhir-jpaserver-starter/pull/8/files#diff-b991193457013b3c05a76d754be6bf2ee7db702ca5cfd879784a63073bb86f96R85-R87

Comment (jdkizer9) [reply src/main/java/ca/uhn/fhir/jpa/tcp/interceptor/TokenAuthorizationInterceptor.java:88]: I'm not sure I follow. The scenario is that someone makes a request for `QuestionnaireResponse` resources, specifying a patient ID. How do we ensure that the server only returns resources associated with the authenticated user? Also, if one were to reference a resource by id (e.g., `QuestionniareResponse/1234`), how do we ensure that the authenticated user has permissions to view that?

File: src/main/java/ca/uhn/fhir/jpa/tcp/service/TcpResourceService.java [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+import java.util.UUID;
+import java.util.stream.Collectors;
+
+@Service
+public class TcpResourceService implements ITcpResourceService {
+
+  public static final String COMMON_HEALTH_ORG_NAME = "CommonHealth";
+  public static final String COMMON_HEALTH_PATIENT_RESOURCE = "PATIENT";
+  public static final String COMMON_HEALTH_PRACTITIONER_RESOURCE = "PRACTITIONER";
+  public static final String COMMON_HEALTH_EID_SYSTEM = "https://auth.tcpdev.org/identity/restv1";
```
Comment (misterzero) [inline src/main/java/ca/uhn/fhir/jpa/tcp/service/TcpResourceService.java:32]: This should be externalized, maybe add to `EmpiConfig`?

Comment (marcovilla86) [reply src/main/java/ca/uhn/fhir/jpa/tcp/service/TcpResourceService.java:32]: ok

File: src/main/java/ca/uhn/fhir/jpa/tcp/interceptor/TokenAuthorizationInterceptor.java [hunk 4]
```diff
@@ -1,93 +1,124 @@
...
 
 @Service
 public class TokenAuthorizationInterceptor extends AuthorizationInterceptor {
-	@Autowired
-	private Environment env;
+  @Autowired
+  private Environment env;
+  @Autowired
+  private ITcpResourceService tcpResourceService;
+  private static final List<String> ALREADY_SCOPED_WORDS = Arrays.asList(new String[]{"patient/*.write", "QuestionnaireResponse",
```
Comment (misterzero) [inline src/main/java/ca/uhn/fhir/jpa/tcp/interceptor/TokenAuthorizationInterceptor.java:39]: This should be loaded from config.

**Review by misterzero — APPROVED** [review]

**Pull request merged**